data-generation/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+    "insurance",
+    "investments-and-savings",
+    "expenses-and-other",
+    "cars-and-transportation",
+];
+
+/// We have different channels
+/// We will choose mobile with probability 0.25 classic otherwise
+const CHANNELS: (&str, &str) = ("mobile", "web");
+
+/// Generate data for a single customer;
+/// Returns expected_amnt of rows in a fixed schema.
+fn generate_customer_transactions(
+    id: i64,
+    expected_amnt: i64,
+    distr_trx: Uniform<f64>,
+    distr_cat: Uniform<i64>,
+    seed: u64,
+    d_minus: i64,
+) -> (
+    Vec<i64>,
+    Vec<&'static str>,
+    Vec<&'static str>,
+    Vec<&'static str>,
+    Vec<f64>,
+    Vec<i64>,
+) {
+    let mut rng = ChaCha8Rng::seed_from_u64(seed);
+
+    // We will sample from expected amount;
+    // Interval is defined on the top of the file
+    let distr_90p = Uniform::<i64>::try_from((expected_amnt * BOTTOM / 100)..(expected_amnt * UPPER / 100)).unwrap();
+    let sampled_expected = distr_90p.sample(&mut rng);
+
+    // I will store rows as columns in form of vectors
+    let ids: Vec<i64> = vec![id; sampled_expected.as_usize()];
+    let mut card_t: Vec<&'static str> = Vec::with_capacity(sampled_expected.as_usize());
+    let mut trx_t: Vec<&'static str> = Vec::with_capacity(sampled_expected.as_usize());
+    let mut ch_t: Vec<&'static str> = Vec::with_capacity(sampled_expected.as_usize());
+    let mut amnt: Vec<f64> = Vec::with_capacity(sampled_expected.as_usize());
+    let t_minus: Vec<i64> = vec![d_minus; sampled_expected.as_usize()];
+
+
+    for _i in 0..sampled_expected {
+        // Choose DC with probability 75%
+        let card_type = if rng.gen_bool(0.75) {
+            CARD_TYPES.0
+        } else {
+            CARD_TYPES.1
+        };
+
+        // Choose random transaction type
+        let trx_type = TRANSACTION_TYPES
+            .get(distr_cat.sample(&mut rng).as_usize())
+            .unwrap();
+
+        // Choose mobile channel with probability 25%
+        let channel_type = if rng.gen_bool(0.25) {
+            CHANNELS.0
+        } else {
+            CHANNELS.1
+        };
+
+        // Sample from uniform distribution for an amount
+        let trx_amnt = distr_trx.sample(&mut rng);
+
+        card_t.push(card_type);
+        trx_t.push(trx_type);
+        ch_t.push(channel_type);
+        amnt.push(trx_amnt);
+    }
+
+    (ids, card_t, trx_t, ch_t, amnt, t_minus)
+}
+
+/// Generate batch of random seeded transactions for the given customers list considering expected amount of transactions
+#[pyfunction]
+fn generate_data_batch(
+    ids: PyArrowType<ArrayData>,
+    trx_per_day: PyArrowType<ArrayData>,
+    days_in_batch: i64,
+    offset: i64,
+    global_seed: u64,
+    partition_name: &str,
+) -> PyArrowType<RecordBatch> {
+    let ids_arr = make_array(ids.0);
+    let expected_arr = make_array(trx_per_day.0);
+    let dim = ids_arr.len() * days_in_batch.as_usize();
+
+    let trx_amnt_distr = Uniform::<f64>::try_from(MIN..MAX).unwrap();
```

Comment:
As an option, consider returning Result<PyArrowType<RecordBatch>>, all `unwrap` will become `?`. Also, `expect` is like unwrap but at least with a message.

Reply:
The problem here is that I exactly know that this code is valid. But there is no other constructor... I will think about it later. I left it as is for now. Thank you anyway!